prettyprinter/src/Data/Text/Prettyprint/Doc/Internal.hs

```diff
@@ -540,35 +541,55 @@ group x = case changesUponFlattening x of
 -- See https://github.com/quchen/prettyprinter/issues/22 for the  corresponding
 -- ticket.
 
+data FlattenResult a
+
+    -- | a is likely flatter than the input.
+    = Flattened a
+
+    -- | The input was already flat, e.g. a 'Text'.
+    | AlreadyFlat
+
+    -- | The input couldn't be flattened: It contained a 'Line' or 'Fail'.
+    | NeverFlat
+
+instance Functor FlattenResult where
+    fmap f (Flattened a) = Flattened (f a)
+    fmap _ AlreadyFlat   = AlreadyFlat
+    fmap _ NeverFlat     = NeverFlat
+
 -- | Choose the first element of each @Union@, and discard the first field of
 -- all @FlatAlt@s.
 --
--- The result is 'Just' if the element might change depending on the layout
--- algorithm (i.e. contains differently renderable sub-documents), and 'Nothing'
--- if the document is static (e.g. contains only a plain 'Empty' node). See
--- [Group: special flattening] for further explanations.
-changesUponFlattening :: Doc ann -> Maybe (Doc ann)
+-- The result is 'Flattened' if the element might change depending on the layout
+-- algorithm (i.e. contains differently renderable sub-documents), and 'AlreadyFlat'
+-- if the document is static (e.g. contains only a plain 'Empty' node).
+-- 'NeverFlat' is returned when the document cannot be flattened because it
+-- contains a hard 'Line' or 'Fail'.
+-- See [Group: special flattening] for further explanations.
+changesUponFlattening :: Doc ann -> FlattenResult (Doc ann)
 changesUponFlattening = \doc -> case doc of
-    FlatAlt _ y     -> Just (flatten y)
-    Line            -> Just Fail
-    Union x _       -> changesUponFlattening x <|> Just x
+    FlatAlt _ y     -> Flattened (flatten y)
+    Line            -> NeverFlat
+    Union x _       -> Flattened x
```

Comment:
This also includes #100 / fixes #99, mostly because I couldn't be bothered to write an `Alternative` instance.